### PR TITLE
[block#LastResultsHash] add Events + GasWanted/Used

### LINKED ISSUE #1007: Add events to results hash
Should be in: https://github.com/tendermint/tendermint/pull/999

But waiting on possible changes to abci responses: https://github.com/tendermint/abci/issues/165

### PR DESCRIPTION
**Pull request by melekes**
Closes #1007 

Spec: https://github.com/tendermint/spec/pull/97/files

### CODE REVIEW

**Comment by melekes** [comment]
I can work more on tests probably. Other than that, this PR is ready.

**Comment by marbar3778** [comment]
it looks like you were blocked on state encoding migration. this has been merged into master. Do you need anything else to help unblock this?

**Comment by melekes** [comment]
> it looks like you were blocked on state encoding migration. this has been merged into master. Do you need anything else to help unblock this?

Cool, thanks! I will update this PR tomorrow

File: state/store.go [hunk 1]
```diff
@@ -253,13 +254,37 @@ func PruneStates(db dbm.DB, from int64, to int64) error {
...
+
+// ABCIResponsesResultsHash returns the root hash of a Merkle tree with 3 leafs:
+//   1) amino encoded ResponseBeginBlock.Events
+//   2) root hash of a Merkle tree of ResponseDeliverTx responses (see ABCIResults.Hash)
+//   3) amino encoded ResponseEndBlock.Events
+//
+// See merkle.SimpleHashFromByteSlices
+func ABCIResponsesResultsHash(ar *tmstate.ABCIResponses) []byte {
+	// Amino-encode BeginBlock events.
+	bbeBytes, err := cdc.MarshalBinaryLengthPrefixed(ar.BeginBlock.Events)
```
Comment (melekes) [inline state/store.go:267]: TODO: use `proto.Marshal` here

File: state/store.go [hunk 2]
```diff
@@ -253,13 +254,37 @@ func PruneStates(db dbm.DB, from int64, to int64) error {
...
+	if err != nil {
+		panic(err)
+	}
+
+	// Build a Merkle tree of amino-encoded DeliverTx results and get a hash.
 	results := types.NewResults(ar.DeliverTxs)
-	return results.Hash()
+
+	// Amino-encode EndBlock events.
+	ebeBytes, err := cdc.MarshalBinaryLengthPrefixed(ar.EndBlock.Events)
```
Comment (melekes) [inline state/store.go:276]: TODO: use `proto.Marshal` here

File: scripts/protocgen.sh [hunk 3]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env bash
 
-set -eo pipefail
+set -e pipefail
```
Comment (melekes) [inline scripts/protocgen.sh:3]:
```suggestion
set -eo pipefail
```

File: state/store.go [hunk 4]
```diff
@@ -253,13 +254,41 @@ func PruneStates(db dbm.DB, from int64, to int64) error {
...
+	if err != nil {
+		panic(err)
+	}
+
+	// Build a Merkle tree of amino-encoded DeliverTx results and get a hash.
 	results := types.NewResults(ar.DeliverTxs)
-	return results.Hash()
+
+	// proto-encode EndBlock events.
+	ebeBytes, err := proto.Marshal(&abci.ResponseEndBlock{
```
Comment (melekes) [inline state/store.go:278]: not sure if it's the best way, but kinda works

File: state/store.go [hunk 5]
```diff
@@ -253,13 +254,41 @@ func PruneStates(db dbm.DB, from int64, to int64) error {
...
+func ABCIResponsesResultsHash(ar *tmstate.ABCIResponses) []byte {
+	// proto-encode BeginBlock events.
+	bbeBytes, err := proto.Marshal(&abci.ResponseBeginBlock{
+		Events: ar.BeginBlock.Events,
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	// Build a Merkle tree of amino-encoded DeliverTx results and get a hash.
```
Comment (marbar3778) [inline state/store.go:274]:
```suggestion
	// Build a Merkle tree of proto-encoded DeliverTx results and get a hash.
```
?

File: state/store.go (hunk 4 above)
Comment (marbar3778) [reply state/store.go:278]: this looks good.

Comment (melekes) [reply state/store.go:278]: you mean `X.Marshal()`?

**Comment by melekes** [comment]
>  Do you think there is a need to have this as a proto type. do we ever marshal this type?

ABCIResults - no. it's a deterministic component of ResponseDeliverTx essentially.

ABCIResponses - yes. I think we save it to the state/store.

File: state/store.go (hunk 4 above)
Comment (marbar3778) [reply state/store.go:278]: yes

**Comment by marbar3778** [comment]
is this ready to be merged? There are some changes that i am duplicating in another branch but would rather just merge this

**Pull request merged**